File: cordova-demo/cordova-sqlite-demo-plugin/src/ios/SQLiteDemo.m
#import <Cordova/CDVPlugin.h>

#include "sqlite-connection-core.h"

@interface SQLiteDemo : CDVPlugin

- (void) openDatabaseConnection: (CDVInvokedUrlCommand *) commandInfo;

- (void) executeBatch: (CDVInvokedUrlCommand *) commandInfo;

@end

@implementation SQLiteDemo

- (void) pluginInitialize
{
  scc_init();
}

- (void) openDatabaseConnection: (CDVInvokedUrlCommand *) commandInfo
{
  NSArray * _args = commandInfo.arguments;

  NSDictionary * options = (NSDictionary *)[_args objectAtIndex: 0];

  const char * filename = [(NSString *)[options valueForKey: @"path"] cString];

  const int flags = [(NSNumber *)[options valueForKey: @"flags"] intValue];

  const int connection_id = scc_open_connection(filename, flags);

  if (connection_id < 0) {
    CDVPluginResult * openErrorResult =
      [CDVPluginResult resultWithStatus: CDVCommandStatus_ERROR
                        messageAsString: @"open error"];
    [self.commandDelegate sendPluginResult: openErrorResult
                                callbackId: commandInfo.callbackId];
    return;
  }

  CDVPluginResult * openResult =
    [CDVPluginResult resultWithStatus: CDVCommandStatus_OK
                         messageAsInt: connection_id];

  [self.commandDelegate sendPluginResult: openResult
                              callbackId: commandInfo.callbackId];
}

- (void) executeBatch: (CDVInvokedUrlCommand *) commandInfo
{
  NSArray * _args = commandInfo.arguments;

  const int connection_id = [(NSNumber *)[_args objectAtIndex: 0] intValue];

  NSArray * data = [_args objectAtIndex: 1];

  NSMutableArray * ra = [NSMutableArray arrayWithCapacity: 0];

  for (int ii=0; ii < [data count]; ++ii) {
    NSArray * saa = [data objectAtIndex:ii];

    NSString * st = [saa objectAtIndex: 0];

    NSArray * aa = [saa objectAtIndex: 1];

    const int rc1 = scc_begin_statement(connection_id, [st cString]);

    int bindResult = 0;

    for (int i=0; i < [aa count]; ++i) {
      NSObject * bindValue = [aa objectAtIndex: i];

      if (bindValue == nil) {
        bindResult = scc_bind_null(connection_id, 1 + i);
      } else if ([bindValue isKindOfClass: [NSNumber class]]) {
        // TBD UIWebView vs WKWebView
        if ([(NSNumber *)bindValue objCType][0] == 'q') {
          bindResult = scc_bind_long(connection_id, 1 + i, [(NSNumber *)bindValue longValue]);
        } else {
          bindResult = scc_bind_double(connection_id, 1 + i, [(NSNumber *)bindValue doubleValue]);
        }
      } else if ([bindValue isKindOfClass: [NSString class]]) {
        bindResult = scc_bind_text(connection_id, 1 + i, [(NSString *)bindValue cString]);
      } else {
        bindResult = scc_bind_null(connection_id, 1 + i);
      }
    }

    if (rc1 != 0 || bindResult != 0) {
      [ra addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];

      scc_end_statement(connection_id);

      continue;
    }

    int rc2 = scc_step(connection_id);

    if (rc2 == 100) {
      const int cc = scc_get_column_count(connection_id);

      NSMutableArray * columns = [NSMutableArray arrayWithCapacity: 0];

      for (int ci = 0; ci < cc; ++ci) {
        NSString * columnNameAsString =
          [NSString stringWithUTF8String: scc_get_column_name(connection_id, ci)];
        [columns addObject: columnNameAsString];
      }

      NSMutableArray * rra = [NSMutableArray arrayWithCapacity: 0];

      int nextrc;

      do {
        NSMutableArray * row = [NSMutableArray arrayWithCapacity: 0];

        for (int c = 0; c < cc; ++c) {
          const int columnType = scc_get_column_type(connection_id, c);

          if (columnType == SCC_COLUMN_TYPE_NULL) {
            [row addObject: [NSNull null]];
          } else if (columnType == SCC_COLUMN_TYPE_INTEGER) {
            NSNumber * columnNumberValue =
              [NSNumber numberWithLongLong: scc_get_column_long(connection_id, c)];
            [row addObject: columnNumberValue];
          } else if (columnType == SCC_COLUMN_TYPE_FLOAT) {
            NSNumber * columnNumberValue =
              [NSNumber numberWithDouble: scc_get_column_double(connection_id, c)];
            [row addObject: columnNumberValue];
          } else {
            NSString * columnStringValue =
              [NSString stringWithUTF8String: scc_get_column_text(connection_id, c)];
            [row addObject: columnStringValue];
          }
        }

        [rra addObject: row];

        nextrc = scc_step(connection_id);
      } while(nextrc == 100);

      [ra addObject: @{@"status":@0, @"columns": columns, @"rows": rra}];
    } else if (rc2 == 101) {
      [ra addObject: @{
        @"status": @0,
        @"total_changes": [NSNumber numberWithInteger:
          scc_get_total_changes(connection_id)],
        @"last_insert_rowid": [NSNumber numberWithInteger:
          scc_get_last_insert_rowid(connection_id)]
      }];
    } else {
      [ra addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];
    }

    scc_end_statement(connection_id);
  }

  CDVPluginResult * batchResult =
    [CDVPluginResult resultWithStatus: CDVCommandStatus_OK
                       messageAsArray: ra];

  [self.commandDelegate sendPluginResult: batchResult
                              callbackId: commandInfo.callbackId];
}

@end

File: cordova-demo/cordova-sqlite-demo-plugin/src/android/SQLiteDemo.java
package com.demo;

import io.sqlc.SCCoreGlue;

import org.apache.cordova.*;

import org.json.*;

public class SQLiteDemo extends CordovaPlugin {
  @Override
  public boolean execute(String method, JSONArray data, CallbackContext cbc) {
    switch(method) {
      case "openDatabaseConnection":
        openDatabaseConnection(data, cbc);
        break;
      case "executeBatch":
        executeBatch(data, cbc);
        break;
      default:
        return false;
    }
    return true;
  }

  static private void
  openDatabaseConnection(JSONArray args, CallbackContext cbc) {
    try {
      final JSONObject options = args.getJSONObject(0);

      final String fullName = options.getString("fullName");

      final int flags = options.getInt("flags");

      final int mydbc = SCCoreGlue.scc_open_connection(fullName, flags);

      if (mydbc < 0) {
        cbc.error("open error: " + -mydbc);
      } else {
        cbc.success(mydbc);
      }
    } catch(Exception e) {
      // NOT EXPECTED - internal error:
      cbc.error(e.toString());
    }
  }

  static private void executeBatch(JSONArray args, CallbackContext cbc) {
    try {
      final int mydbc = args.getInt(0);

      JSONArray data = args.getJSONArray(1);

      JSONArray results = new JSONArray();
      final int count = data.length();

      for (int i=0; i<count; ++i) {
        int previousTotalChanges = SCCoreGlue.scc_get_total_changes(mydbc);

        JSONArray entry = data.getJSONArray(i);

        String s = entry.getString(0);

        if (SCCoreGlue.scc_begin_statement(mydbc, s) != 0) {
          JSONObject result = new JSONObject();
          result.put("status", 1); // REPORT SQLite ERROR 1
          result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
          results.put(result);
        } else {
          JSONArray bind = entry.getJSONArray(1);

          final int bindCount = bind.length();

          int bindResult = 0; // SQLite OK

          for (int j = 0; j < bindCount; ++j) {
            final Object o = bind.get(j);

            if (o instanceof Number) {
              bindResult =
                SCCoreGlue.scc_bind_double(mydbc, 1 + j, bind.optDouble(j));
            } else if (o instanceof String) {
              bindResult =
                SCCoreGlue.scc_bind_text(mydbc, 1 + j, o.toString());
            } else {
              bindResult =
                SCCoreGlue.scc_bind_null(mydbc, 1 + j);
            }
          }

          if (bindResult != 0) {
            JSONObject result = new JSONObject();
            result.put("status", 1); // REPORT SQLite ERROR 1
            result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
            continue;
          }

          int stepResult = SCCoreGlue.scc_step(mydbc);

          if (stepResult == 100) {
            final int columnCount = SCCoreGlue.scc_get_column_count(mydbc);

            JSONArray columns = new JSONArray();

            for (int j=0; j < columnCount; ++j) {
              columns.put(SCCoreGlue.scc_get_column_name(mydbc, j));
            }

            JSONArray rows = new JSONArray();

            do {
              JSONArray row = new JSONArray();

              for (int col=0; col < columnCount; ++col) {
                final int type = SCCoreGlue.scc_get_column_type(mydbc, col);

                if (type == SCCoreGlue.SCC_COLUMN_TYPE_INTEGER ||
                    type == SCCoreGlue.SCC_COLUMN_TYPE_FLOAT) {
                  row.put(SCCoreGlue.scc_get_column_double(mydbc, col));
                } else if (type == SCCoreGlue.SCC_COLUMN_TYPE_NULL) {
                  row.put(JSONObject.NULL);
                } else {
                  row.put(SCCoreGlue.scc_get_column_text(mydbc, col));
                }
              }

              rows.put(row);

              stepResult = SCCoreGlue.scc_step(mydbc);
            } while (stepResult == 100);

            JSONObject result = new JSONObject();
            result.put("status", 0); // REPORT SQLite OK
            result.put("columns", columns);
            result.put("rows", rows);
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          } else if (stepResult == 101) {
            int totalChanges = SCCoreGlue.scc_get_total_changes(mydbc);
            int rowsAffected = totalChanges - previousTotalChanges;

            JSONObject result = new JSONObject();
            // same order as iOS & macOS ("osx"):
            result.put("status", 0); // REPORT SQLite OK
            result.put("totalChanges", totalChanges);
            result.put("rowsAffected", rowsAffected);
            result.put("lastInsertRowId",
              SCCoreGlue.scc_get_last_insert_rowid(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          } else {
            JSONObject result = new JSONObject();
            result.put("status", 1); // REPORT SQLite ERROR 1
            result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          }
        }
      }

      // send results to JavaScript side (...)
      cbc.success(results);
    } catch(Exception e) {
      // NOT EXPECTED - internal error:
      cbc.error(e.toString());
    }
  }

  static {
    System.loadLibrary("sqlite-connection-core-glue");
    SCCoreGlue.scc_init();
  }
}
